moco-switchover-downtime-monitor/checker.go

```diff
@@ -0,0 +1,442 @@
+	c.logger.Info("starting checker")
+	defer c.logger.Info("stopping checker")
+
+	for {
+		rand.Shuffle(len(c.methods), func(i, j int) {
+			c.methods[i], c.methods[j] = c.methods[j], c.methods[i]
+		})
+		for _, method := range c.methods {
+			c.cooldown(ctx)
+			if ctx.Err() != nil {
+				return
+			}
+			c.check(ctx, method)
+			if ctx.Err() != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *Checker) cooldown(ctx context.Context) {
+	cooldownDuration := cooldownDurationMin + time.Duration(rand.Int63n(int64(cooldownDurationMax-cooldownDurationMin)))
+	c.logger.Info("begin cooldown", zap.Stringer("duration", cooldownDuration))
+	cooldownTimer := time.NewTimer(cooldownDuration)
+	defer cooldownTimer.Stop()
+	select {
+	case <-ctx.Done():
+		c.logger.Info("cancel cooldown")
+		return
+	case <-cooldownTimer.C:
+	}
+	c.logger.Info("end cooldown")
+}
+
+func (c *Checker) check(ctx context.Context, method Method) {
+	logger := c.logger.With(zap.String("method", method.MethodName()))
+	logger.Info("begin checking")
+
+	c.waitForPreCondition(ctx, logger, method)
+	if ctx.Err() != nil {
+		logger.Info("cancel checking")
+		return
+	}
+
+	timedCtx, cancel := context.WithTimeout(ctx, completionTimeout)
+	defer cancel()
+	pingerCtx, stopPinger := context.WithCancel(timedCtx)
+	defer stopPinger()
+
+	pingerCancelled := false
+	cancelPinger := func() {
+		pingerCancelled = true
+		stopPinger()
+	}
+	pingerCancelledFunc := func() bool {
+		return ctx.Err() != nil || pingerCancelled
+	}
+
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.executeAndWaitForCompletion(ctx, timedCtx, logger, cancelPinger, stopPinger, method)
+	}()
+
+	pingerParams := []struct {
+		Suffix  string
+		DoWrite bool
+	}{
+		{
+			Suffix:  "all",
+			DoWrite: false,
+		},
+		{
+			Suffix:  "primary",
+			DoWrite: true,
+		},
+		{
+			Suffix:  "primary",
+			DoWrite: false,
+		},
+		{
+			Suffix:  "replica",
+			DoWrite: false,
+		},
+	}
+	for _, param := range pingerParams {
+		param := param
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			c.pinger(pingerCtx, logger, pingerCancelledFunc, method.MethodName(), param.Suffix, param.DoWrite)
+		}()
+	}
+
+	wg.Wait()
+
+	if ctx.Err() != nil {
+		logger.Info("cancel checking")
+		return
+	}
+
+	logger.Info("end checking")
+}
+
+func (c *Checker) waitForPreCondition(ctx context.Context, logger *zap.Logger, method Method) {
+	startTime := time.Now()
+	preconditionCheckTicker := time.NewTicker(preconditionCheckInterval)
+	defer preconditionCheckTicker.Stop()
+	preconditionWarningTicker := time.NewTicker(preconditionWarningInterval)
+	defer preconditionWarningTicker.Stop()
+	for {
+		met, err := method.CheckPreCondition(ctx)
+		if err != nil {
+			// cancelled
+			return
+		}
+		if met {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			// cancelled
+			return
+		case <-preconditionWarningTicker.C:
+			logger.Warn("precondition is not met yet", zap.Stringer("duration", time.Since(startTime).Round(time.Second)))
```

Comment:
It may or may not be an abnormal condition.

In our k8s cluster, nodes are occasionally rebooted and MySQLClusters become unhealthy for tens of minutes.